sql/settype.go

```diff
@@ -0,0 +1,359 @@
+	switch value := v.(type) {
+	case int:
+		return t.ConvertToBits(uint64(value))
+	case uint:
+		return t.ConvertToBits(uint64(value))
+	case int8:
+		return t.ConvertToBits(uint64(value))
+	case uint8:
+		return t.ConvertToBits(uint64(value))
+	case int16:
+		return t.ConvertToBits(uint64(value))
+	case uint16:
+		return t.ConvertToBits(uint64(value))
+	case int32:
+		return t.ConvertToBits(uint64(value))
+	case uint32:
+		return t.ConvertToBits(uint64(value))
+	case int64:
+		return t.ConvertToBits(uint64(value))
+	case uint64:
+		if value < t.upperBound() {
+			return value, nil
+		}
+	case float32:
+		return t.ConvertToBits(uint64(value))
+	case float64:
+		return t.ConvertToBits(uint64(value))
+	case string:
+		bitArray, err := t.ConvertStringToBitArray(value)
+		if err != nil {
+			return uint64(0), err
+		}
+		return t.sumBitArray(bitArray), nil
+	case []byte:
+		return t.ConvertToBits(string(value))
+	}
+
+	return uint64(0), ErrConvertingToSet.New(v)
+}
+
+// Promote implements the Type interface.
+func (t setType) Promote() Type {
+	return t
+}
+
+// SQL implements Type interface.
+func (t setType) SQL(v interface{}) (sqltypes.Value, error) {
+	if v == nil {
+		return sqltypes.NULL, nil
+	}
+	value, err := t.Convert(v)
+	if err != nil {
+		return sqltypes.Value{}, err
+	}
+	return sqltypes.MakeTrusted(sqltypes.Set, []byte(value.(string))), nil
+}
+
+// String implements Type interface.
+func (t setType) String() string {
+	s := fmt.Sprintf("SET('%v')", strings.Join(t.Values(), ","))
+	if t.CharacterSet() != Collation_Default.CharacterSet() {
+		s += " CHARACTER SET " + t.CharacterSet().String()
+	}
+	if t.collation != Collation_Default {
+		s += " COLLATE " + t.collation.String()
+	}
+	return s
+}
+
+// Type implements Type interface.
+func (t setType) Type() query.Type {
+	return sqltypes.Set
+}
+
+// Zero implements Type interface.
+func (t setType) Zero() interface{} {
+	return ""
+}
+
+func (t setType) CharacterSet() CharacterSet {
+	return t.collation.CharacterSet()
+}
+
+func (t setType) Collation() Collation {
+	return t.collation
+}
+
+// ConvertBitArrayToString converts the given bit array into the equivalent comma-delimited string.
+// The given bit array must have a form where representing the value 12 would have values {4,8}.
+func (t setType) ConvertBitArrayToString(bitArray []uint64) (string, error) {
+	strBuilder := strings.Builder{}
+	for i, bit := range bitArray {
+		val, ok := t.bitToVal[bit]
+		if !ok {
+			return "", ErrInvalidSetValue.New(t.sumBitArray(bitArray))
+		}
+		if i != 0 {
+			strBuilder.WriteByte(',')
+		}
+		strBuilder.WriteString(val)
+	}
+	return strBuilder.String(), nil
+}
+
+// ConvertStringToBitArray converts the given string into a bit array.
+// Ignores duplicate set values and always returns the bits in ascending order.
+// The returned bit array representing the value 12 will have values {4,8}.
+func (t setType) ConvertStringToBitArray(str string) ([]uint64, error) {
```

Comment:
Rather than returning an array here, you should just use the original bitmask semantics of the set type.